README.md
## About [![build status](https://secure.travis-ci.org/SaltwaterC/dynamo-backup-streams.png?branch=master)](https://travis-ci.org/SaltwaterC/dynamo-backup-streams) [![NPM version](https://badge.fury.io/js/dynamo-backup-streams.png)](http://badge.fury.io/js/dynamo-backup-streams)

Backup / Restore DynamoDB node.js streams. Designed to be used in various pipelines in order to build custom backup / restore solutions. Both Backup and Restore handle properly few things, like back-pressure and capacity usage (including auto-scaling).

## System requirements

 * node.js 6+

`aws-sdk` must be installed before using this library. A DynamoDB client instance must be passed to the Backup / Restore constructor.

## Backup

Implemented as node.js Readable stream. Built on top of DynamoDB Scan with ConsistentRead. It outputs a stream of JSON encoded DynamoDB items separated by newline.

The backup.js example shows how a pipeline with Backup works. In this particular case, the Backup stream is piped into gzip, then piped into a file.

```javascript
var AWS = require('aws-sdk');
var Backup = require('dynamo-backup-streams').Backup;
var backupStream = new Backup({
  client: new AWS.DynamoDB(),
  table: 'table_name'
});

// get an output stream somehow
backupStream.pipe(outputStream); // do stuff with the stream
```

The Backup constructor accepts an options object. The spec for that object:

 * `client` - DynamoDB client instance created with `new AWS.DynamoDB()`
 * `table` - indicates the DynamoDB table name to backup. This is the only compulsory option.
 * `capacityPercentage` - the percentage of the read capacity units to use on DynamoDB to be used as the total target capacity. Default: 25.
 * `retries` - the number of retries for each read request. Default: 10.
 * `delay` - the target delay in milliseconds between each request. Default: 1000.
 * `concurrency` - the number of workers to use for reading the data. Each worker reads a Scan segment. Default: 4.
 * `fixedCapacity` - Number to operate on a fixed capacity e.g. needed to utilize tables in [on demand mode](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadWriteCapacityMode.html#HowItWorks.OnDemand
).
 Make sure to consult the [available throughput in your region](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#default-limits-throughput). Default undefined

The actual capacity requirements are calculated on each request. One read unit equals to 4KB/s of throughput. DynamoDB Scan counts the total size of the items for the purpose of capacity usage which makes this operation very efficient. On each request, the used capacity is compared against the target capacity of the worker. The target capacity is then adjusted accordingly. The total target capacity is checked against DynamoDB every 60 seconds to see if auto-scaling has kicked in. The target capacity of the worker is calculated by dividing the total target capacity to the number of workers and rounding down to the nearest integer unit. The read capacity per worker can't be smaller than 1.

If the size of a typical item is greater than the maximum request size which can be as low as 1 read unit per worker, then the capacity usage would be higher than the specified percentage. You may need to reduce the number of workers to fit into the capacity target.

The maximum throughput which can be achieved per worker is 257 read units (1028 KB/s). This is the throughput limit of DynamoDB partitions. Basically, a worker is needed for every 257 read units in order to fully utilise the read capacity, if desired.

The delay option should not be adjusted, unless there's a reason to do so. Bear in mind that the delay value is a target, not the actual delay between the current and the next request. The capacity is computed dynamically, however, since it's in KB/s, a request needs to be executed, on average, every second. With network latency which can't be predicted and auto-scaling, a fixed offset won't work for achieving this 1 request/second target. Every request is measured to see how much time it takes to process. This time difference is then subtracted from the delay target, then the next request is scheduled based on this dynamic offset.

If the stream is paused, no data is being read from DynamoDB. The status of the paused state is checked every 250 milliseconds. The stream would emit data events for requests still in flight. The stream would be fully paused once all the DynamoDB Scan requests have been completed.

## Restore

Implemented as node.js PassThrough stream. Built on top of DynamoDB BatchWriteItem. The input is a stream created by Backup. For all intents and purposes this object must be used as a Writable stream. The Readable bit is passed internally to [readline](https://nodejs.org/api/readline.html) to tokenise the input stream into lines which then are parsed as JSON. readline is handling the back-pressure whenever the line buffer is filled.

The restore.js example shows how a pipeline with Restore works. In this particular case, a file read stream is piped into gunzip, then piped into the Restore stream.

```javascript
var AWS = require('aws-sdk');
var Restore = require('dynamo-backup-streams').Restore;
var restoreStream = new Restore({
  client: new AWS.DynamoDB(),
  table: 'table_name'
});

// get an input stream somehow
inputStream.pipe(restoreStream);
```

The Restore constructor accepts an options object. The spec for that object:

 * `client` - DynamoDB client instance created with `new AWS.DynamoDB()`
 * `table` - indicates the DynamoDB table name to backup. This is the only compulsory option.
 * `capacityPercentage` - the percentage of write capacity units to use on DynamoDB to be used as the total target capacity. Default: 25.
 * `retries` - the number of retries for each write request. Default: 10.
 * `delay` - the target delay in milliseconds between each request. Default: 1000.
 * `concurrency` - the number of workers to use for writing the data. Default: 1.
 * `bufferSize` - the size of the line buffer created by reading from readline. Default: 250 * `concurrency`.
 * `maxItems` - the maximum number of items in a batchWriteItem request. Default: 25. Max: 25.
 * `fixedCapacity` - Number to operate on a fixed capacity e.g. needed to utilize tables in [on demand mode](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadWriteCapacityMode.html#HowItWorks.OnDemand
).
 Make sure to consult the [available throughput in your region](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#default-limits-throughput). Default undefined

The same formulas for capacity sizing and request timings used for the Backup stream apply. There are certain differences though. One write unit equals to 1KB/s of throughput. DynamoDB BatchWriteItem counts the size of each individual item and the value is rounded up to the nearest KB. This makes the capacity sizing tad inefficient compared to Scan. The general limits for BatchWriteItem apply: up to 25 items per batch, 400KB as the maximum size of the item, and 16MB as the total size of the BatchWriteItem request.

If the lines buffer is full, readline is paused until the buffer size drops under the set limit. Because readline doesn't pause the stream immediately, i.e the in flight input Buffers may still trigger line events, the actual size of the line buffer may be higher, but not by much. By default, the capacity of this buffer is set dynamically. Each worker adds 250 units to the line buffer limit. Since the main limitation is the slow throughput of DynamoDB writes, this buffer would be pretty much full for the whole duration of the restore process.

This stream emits the close event when the line buffer is drained and the readline stream is closed. This indicates that no more writes are being done to DynamoDB.

## AWS Authentication

By default, if no authentication options are being specified, aws-sdk tries to use the environment variables or the instance profile credentials (if running on EC2).

The minimum environment variables required to make it work:

 * `AWS_REGION`
 * `AWS_ACCESS_KEY_ID`
 * `AWS_SECRET_ACCESS_KEY`

To use a credentials profile saved as ini files with environment variables:

 * `AWS_REGION`
 * `AWS_PROFILE`

To specify in code a credentials profile:

```javascript
new AWS.DynamoDB({
  credentials: new AWS.SharedIniFileCredentials({
    profile: 'profile_name'
  })
});
```

## Screenshots

These screenshots display the capacity usage for a large table backup with auto-scaling turned on.

![large-table-1h.png](/screenshots/large-table-1h.png?raw=true "large-table-1h.png")

![large-table-3h.png](/screenshots/large-table-3h.png?raw=true "large-table-3h.png")


CHANGELOG.md
## v0.2.3
 * Merge [#5 - Adding explicit capacity value for Backup / Restore](https://github.com/SaltwaterC/dynamo-backup-streams/pull/5) - the purpose is support DynamoDB on demand mode, although it may be used for any use case where you want a fix capacity rather than depending on the dynamic capacity sizing based on the read values and a supplied percentage.

## v0.2.2
 * Merge [#3 - Add support for creating streams that handle a subset of the total segments](https://github.com/SaltwaterC/dynamo-backup-streams/pull/3)
 * Merge [#4 - Fix bad this reference, use self instead](https://github.com/SaltwaterC/dynamo-backup-streams/pull/4)

## v0.2.1
 * Fix [#1](https://github.com/SaltwaterC/dynamo-backup-streams/issues/1) - `Restore._streamRecords` attempted to write empty batch as it failed to return when `records.length === 0`.

## v0.2.0
 * Decouple aws-sdk from this library.

## v0.1.0
 * Initial release featuring Backup and Restore streams
Fix bad this reference, use self instead


test/backup.js
'use strict';

var fs = require('fs');
var zlib = require('zlib');

var Backup = require('../lib/main').Backup;

var table = process.argv[2];

var bs = new Backup({
  table: table,
  capacityPercentage: 100,
  concurrency: 2
});
var gzip = zlib.createGzip({
  level: 9
});
var ws = fs.createWriteStream(table + '.json.gz');
// var ws = fs.createWriteStream(table + '.json');

bs.pipe(gzip).pipe(ws);
// bs.pipe(ws);

// outputs a lot less noise compared to listening for 'progress' events
var progress = setInterval(function() {
  var percent = 0;
  if (bs.itemsCount > 0 && bs.itemsProcessed > 0) {
    percent = bs.itemsProcessed / bs.itemsCount * 100;
    if (percent > 100) {
      percent = 100;
    }

    if (percent === 100) {
      clearInterval(progress);
    }
  }

  console.log(table + ' progress: %f% - %d of %d; capacity %d - used: %d / %d', percent, bs.itemsProcessed, bs.itemsCount, bs.limit, bs.units.reduce(function(a, b) { return a + b; }, 0), bs.concurrency);
}, 1000);

bs.on('end', function() {
  clearInterval(progress);
});


test/restore.js
'use strict';

var fs = require('fs');
var zlib = require('zlib');
var ps = require('progress-stream');

var Restore = require('../lib/main').Restore;

var table = process.argv[2];

fs.stat(table + '.json.gz', function(err, stat) {
  if(err) {
    throw err;
  }

  var rs = fs.createReadStream(table + '.json.gz');
  var str = ps({
    length: stat.size,
    time: 1000
  });
  var gunzip = zlib.createGunzip();
  var rss = new Restore({
    table: table,
    capacityPercentage: 100
  });

  rs.pipe(str).pipe(gunzip).pipe(rss);

  var progress = setInterval(function() {
    var stats = str.progress();
    console.log(table + ' progress: %d / %d - %f%; read lines: %d - processed lines: %d; capacity: %d - used: %d', stats.transferred, stats.length, stats.percentage, rss.lines, rss.processedLines, rss.limit, rss.units);
  }, 1000);

  rs.on('close', function() {
    cleanInterval(progress);
  });
});

process.on('exit', function() {
  console.log();
  console.log('total lines: %d', rss.lines);
  console.log('processed lines: %d', rss.processedLines);
});


test/backup.test.js
'use strict';

var crypto = require('crypto');
var readline = require('readline');

var AWS = require('aws-sdk');
var assert = require('chai').assert;

var Backup = require('..').Backup;

describe('DynamoDB Backup tests', function() {
  var count = process.env.BACKUP_RECORDS | 0 || 3;
  var records = [],
    ddb = new AWS.DynamoDB();
  var nodeMajor = process.versions.node.split('.')[0];
  var table = process.env.BACKUP_TABLE + nodeMajor;

  before(function(done) {
    var idx, payload, record, putItems = [];

    for (idx = 0; idx < count; idx++) {
      payload = Date.now().toString() + idx;

      record = {
        payload: {
          S: crypto.createHash('sha256').update(payload).digest('hex')
        },
        primaryKey: {
          N: idx.toString()
        }
      };
      records.push(record);
      putItems.push({
        PutRequest: {
          Item: record
        }
      });
    }

    var reqItems = {
      RequestItems: {}
    };
    reqItems.RequestItems[table] = putItems;

    ddb.batchWriteItem(reqItems, function(err) {
      assert.ifError(err, 'DynamoDB fails to write items');
      done();
    });
  });

  describe('backup table', function() {
    it('is expected to backup records from table', function(done) {
      var lineCount = 0;

      var backup = new Backup({
        client: ddb,
        table: table,
        concurrency: 1,
        capacityPercentage: 2500
      });

      var lines = readline.createInterface({
        input: backup,
        terminal: false
      });

      lines.on('line', function(line) {
        var idx = JSON.parse(line).primaryKey.N;
        assert.strictEqual(line, JSON.stringify(records[idx]), 'record contents');
        lineCount++;
      });

      lines.on('close', function() {
        assert.strictEqual(lineCount, count, 'number of processed records');
        done();
      });
    });
  });
});


lib/backup.js
'use strict';

var util = require('util');
var Readable = require('stream').Readable;

function Backup(options) {
  this.ddb = options.client;
  this.fixedCapacity = parseInt(options.fixedCapacity);
  this.table = options.table;
  this.capacityPercentage = options.capacityPercentage || 25;
  this.retries = options.retries || 10;
  this.delay = options.delay || 1000;
  this.concurrency = options.concurrency || 4;
  this.startSegment = options.startSegment || 0;
  this.totalSegments = options.totalSegments || this.concurrency;

  this.capacityInterval = null;
  this.itemsCount = 0;
  this.itemsProcessed = 0;
  this.units = [];
  this.finished = 0;

  Readable.call(this, options);

  this._readTable();
}
util.inherits(Backup, Readable);

Backup.prototype._read = function() {
  return 0;
};

Backup.prototype._end = function() {
  this.finished++;

  if (this.finished === this.concurrency) {
    this.push(null);
    clearInterval(this.capacityInterval);
    this.emit('close');
  }
};

Backup.prototype._setCapacity = function(units) {
  if (!Number.isNaN(this.fixedCapacity)) {
    units = this.fixedCapacity;
  }
  var readPercentage = (units * this.capacityPercentage / 100) | 0;
  this.limit = Math.max(readPercentage, 1);
};

Backup.prototype._readTable = function() {
  var self = this;

  this.ddb.describeTable({
    TableName: this.table
  }, function(err, data) {
    if (err) {
      return self.emit('error', err);
    }

    self.itemsCount = data.Table.ItemCount;
    self._setCapacity(data.Table.ProvisionedThroughput.ReadCapacityUnits);

    var limit = self.limit / self.concurrency | 0;
    limit = Math.max(limit, 1);

    var segment = self.startSegment;
    for (segment; segment < self.startSegment + self.concurrency; segment++) {
      self._streamRecords(null, segment, limit, 0);
    }
  });

  // update the capacity limit every minute if auto scaling kicks in
  this.capacityInterval = setInterval(function() {
    self.ddb.describeTable({
      TableName: self.table
    }, function(err, data) {
      self._setCapacity(data.Table.ProvisionedThroughput.ReadCapacityUnits);
    });
  }, 60000);
};

Backup.prototype._streamRecords = function(startKey, segment, limit, retries) {
  var self = this;
  var startTime = Date.now();

  // don't buffer data in the stream if the stream is paused
  if (this.isPaused()) {
    return setTimeout(function() {
      self._streamRecords(startKey, segment, limit, 0);
    }, 250);
  }

  var options = {
    TableName: this.table,
    Limit: limit,
    ReturnConsumedCapacity: 'TOTAL',
    ConsistentRead: true,
    Segment: segment,
    TotalSegments: this.totalSegments
  };

  if (startKey) {
    options.ExclusiveStartKey = startKey;
  }

  this.ddb.scan(options, function(err, data) {
    var idx;
    var endTime = Date.now();

    if (err) {
      if (retries >= self.retries) {
        clearInterval(self.capacityInterval);
        var error = new Error('Fail: too many retries when reading records');
        error.dynamoError = err;
        return self.emit('error', error);
      }

      return setTimeout(function() {
        retries++;
        self._streamRecords(startKey, segment, limit, retries);
      }, 1000 * retries);
    }

    if (data.Items.length > 0) {
      for (idx in data.Items) {
        self.push(JSON.stringify(data.Items[idx]) + '\n');
        self.itemsProcessed++;
      }
    }

    // dynamically scale read capacity based on record sizes rather than
    // a limit which is pegged against the number of records returned
    var units = data.ConsumedCapacity.CapacityUnits;
    limit = (limit * self.limit / units) / self.concurrency | 0;
    limit = Math.max(limit, 1);
    self.units[segment] = units;

    self.emit('progress', {
      count: self.itemsCount,
      processed: self.itemsProcessed
    });

    if (!data.LastEvaluatedKey) {
      self._end();
    } else {
      setTimeout(function() {
        self._streamRecords(data.LastEvaluatedKey, segment, limit, 0);
      }, self.delay - (endTime - startTime));
    }
  });
};

module.exports = Backup;


lib/restore.js
'use strict';

var util = require('util');
var readline = require('readline');
var PassThrough = require('stream').PassThrough;

function Restore(options) {
  this.ddb = options.client;
  this.fixedCapacity = parseInt(options.fixedCapacity);
  this.table = options.table;
  this.capacityPercentage = options.capacityPercentage || 25;
  this.retries = options.retries || 10;
  this.delay = options.delay || 1000;
  this.concurrency = options.concurrency || 1;
  this.bufferSize = options.bufferSize || 250 * this.concurrency;
  this.maxItems = options.maxItems || 25;
  this.maxItems = Math.min(this.maxItems, 25);

  this.capacityInterval = null;
  this.lines = 0;
  this.processedLines = 0;
  this.lineBuffer = [];
  this.units = [];
  this.readClosed = false;
  this.dynamoStream = false;
  this.workers = this.concurrency;

  PassThrough.call(this, options);

  this._readLines();
}
util.inherits(Restore, PassThrough);

Restore.prototype._readLines = function() {
  var self = this;

  this.readline = readline.createInterface({
    input: this,
    terminal: false
  });

  this.readline.on('line', function(line) {
    self.lines++;
    self.lineBuffer.push({
      PutRequest: {
        Item: JSON.parse(line)
      }
    });

    if (self.lineBuffer.length > self.bufferSize) {
      self.readline.pause();
    }

    if (!self.dynamoStream) {
      self.dynamoStream = true;
      self._writeTable();
    }
  });

  this.readline.on('close', function() {
    self.readClosed = true;
    clearInterval(self.capacityInterval);
  });

  this.readline.on('error', function(err) {
    self.emit('error', err);
  });
};

Restore.prototype._setCapacity = function(units) {
  if (!Number.isNaN(this.fixedCapacity)) {
    units = this.fixedCapacity;
  }
  var writePercentage = (units * this.capacityPercentage / 100) | 0;
  this.limit = Math.max(writePercentage, 1);
};

Restore.prototype._writeTable = function() {
  var self = this;

  this.ddb.describeTable({
    TableName: this.table
  }, function(err, data) {
    if (err) {
      return self.emit('error', err);
    }

    self._setCapacity(data.Table.ProvisionedThroughput.WriteCapacityUnits);

    var limit = self.limit / self.concurrency | 0;
    limit = Math.max(limit, 1);

    var worker = 0;
    for (worker; worker < self.concurrency; worker++) {
      self._streamRecords(worker, 1, 0, null);
    }
  });

  // update the capacity limit every minute if auto scaling kicks in
  this.capacityInterval = setInterval(function() {
    self.ddb.describeTable({
      TableName: self.table
    }, function(err, data) {
      self._setCapacity(data.Table.ProvisionedThroughput.WriteCapacityUnits);
    });
  }, 60000);
};

Restore.prototype._streamRecords = function(worker, limit, retries, reqItems) {
  var self = this;
  var startTime = Date.now();

  if (this.lineBuffer.length < this.bufferSize) {
    this.readline.resume();
  }

  limit = Math.min(limit, this.maxItems);
  var records = this.lineBuffer.splice(0, limit);

  if (records.length === 0) {
    if (!this.readClosed) {
      this.readline.resume();
      setTimeout(function() {
        self._streamRecords(worker, limit, 0, null);
      }, 1000);
    } else {
      this._close();
    }
    return;
  }

  if (!reqItems) {
    reqItems = {
      ReturnConsumedCapacity: 'TOTAL',
      RequestItems: {}
    };
    reqItems.RequestItems[this.table] = records;
  }

  this.ddb.batchWriteItem(reqItems, function(err, data) {
    var endTime = Date.now();

    if (err) {
      if (retries >= self.retries) {
        clearInterval(self.capacityInterval);
        var error = new Error('Fail: too many retries when writing records');
        error.dynamoError = err;
        return self.emit('error', error);
      }

      return setTimeout(function() {
        retries++;
        self._streamRecords(worker, limit, retries, reqItems);
      }, 1000 * retries);
    }

    // dynamically scale write capacity
    var units = data.ConsumedCapacity[0].CapacityUnits;
    limit = (limit * self.limit / units) / self.concurrency | 0;
    limit = Math.max(limit, 1);
    self.units[worker] = units;

    if (data.UnprocessedItems && data.UnprocessedItems[self.table]) {
      self.processedLines += reqItems.RequestItems[self.table].length -
        data.UnprocessedItems[self.table].length;

      reqItems = {
        ReturnConsumedCapacity: 'TOTAL',
        RequestItems: {}
      };
      reqItems.RequestItems = data.UnprocessedItems;

      setTimeout(function() {
        self._streamRecords(worker, limit, 0, null);
      }, self.delay - (endTime - startTime));
    } else {
      self.processedLines += reqItems.RequestItems[self.table].length;

      if (self.lineBuffer.length > 0) {
        setTimeout(function() {
          self._streamRecords(worker, limit, 0, null);
        }, self.delay - (endTime - startTime));
      } else {
        if (!self.readClosed) {
          // wait for the read buffer to fill
          self.readline.resume();
          setTimeout(function() {
            self._streamRecords(worker, limit, 0, null);
          }, 1000);
        } else {
          self._close();
        }
      }
    }
  });
};

Restore.prototype._close = function() {
  this.workers--;
  if (this.workers === 0) {
    this.emit('close');
  }
};

module.exports = Restore;
